metadata/src/utils/retain.rs

```diff
@@ -237,58 +167,189 @@ pub fn retain_metadata<F, G>(
+    // can still decode values into it).
+    let outer_enums = metadata.outer_enums();
+    let mut find_type_id =
+        collect_return_types(metadata, &mut pallets_filter, &mut runtime_apis_filter);
+    for outer_enum_ty_id in [
+        outer_enums.call_enum_ty(),
+        outer_enums.error_enum_ty(),
+        outer_enums.event_enum_ty(),
+    ] {
+        if !find_type_id(outer_enum_ty_id) {
+            strip_variants_in_enum_type(metadata, &mut pallets_filter, outer_enum_ty_id);
         }
-        should_retain
-    });
+    }
+
+    // 4. Collect all of the type IDs we still want to keep after deleting.
+    let mut keep_these_type_ids: BTreeSet<u32> =
+        iterate_metadata_types(metadata).map(|x| *x).collect();
 
-    // Additionally, subxt depends on the `DispatchError` type existing; we use the same
-    // logic here that is used when building our `Metadata`.
+    // 5. Additionally, subxt depends on the `DispatchError` type existing; we use the same
+    // logic here that is used when building our `Metadata` to ensure we keep it too.
     let dispatch_error_ty = metadata
         .types
         .types
         .iter()
         .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
         .expect("Metadata must contain sp_runtime::DispatchError");
-    type_ids.insert(dispatch_error_ty.id);
 
-    // Now, keep the type IDs we've asked for. This recursively keeps any types referenced from these.
-    // This will return a map from old to new type ID, because IDs may change.
-    let map_ids = metadata.types.retain(|id| type_ids.contains(&id));
+    keep_these_type_ids.insert(dispatch_error_ty.id);
+
+    // 5. Strip all of the type IDs we no longer need, based on the above set.
+    let map_ids = metadata
+        .types
+        .retain(|id| keep_these_type_ids.contains(&id));
+
+    // 6. Now, update the type IDs referenced in our metadata to reflect this.
+    for id in iterate_metadata_types(metadata) {
+        if let Some(new_id) = map_ids.get(id) {
+            *id = *new_id;
+        } else {
+            panic!("Type id {id} was not retained. This is a bug");
+        }
+    }
+}
+
+fn strip_variants_in_enum_type<F>(metadata: &mut Metadata, mut pallets_filter: F, id: u32)
+where
+    F: FnMut(&str) -> bool,
+{
+    let ty = {
+        metadata
+            .types
+            .types
+            .get_mut(id as usize)
+            .expect("Metadata should contain enum type in registry")
+    };
+
+    let TypeDef::Variant(variant) = &mut ty.ty.type_def else {
+        panic!("Metadata type is expected to be a variant type");
+    };
+
+    variant.variants.retain(|v| pallets_filter(&v.name));
+}
+
+/// Returns an iterator that allows modifying each type ID seen in the metadata (not recursively).
+/// This will iterate over every type referenced in the metadata outside of `metadata.types`.
+fn iterate_metadata_types(metadata: &mut Metadata) -> impl Iterator<Item = &mut u32> {
```

Comment:
I wonder whether we can name this function to something better explain that you get a mutable iterator to modify the metadata.

What I can come up with is `iter_mut(&mut Metadata), metadata_types_iter_mut(&mut Metadata)` :shrug: